rfcs/20190225-tf-on-demand.md

```diff
@@ -0,0 +1,191 @@
+
+
+
+*   *Hard Constraint Failures*
+    *   32-bit OS
+    *   GPU too old (Cuda Compute Capability older than 3.0)
+*   *Soft Constraint Failures*
+    *   CUDA version too old
+    *   cuDNN missing
+    *   Other runtime libraries missing or too old (libstdc++)
+    *   Unsupported python version found
+
+In hard constraint failure cases, the tool will point users to cloud options. In soft constraint failure cases, the tool will offer alternative build configurations (CPU only) or request users to install the missing software.
+
+We propose distributing this binary through PyPI. However, please note that binary distribution method is still under discussion.
+
+### Back-end
+
+**TF Builder Service**
+
+TF builder service will require a proto with build options. This proto will define all options the system recognizes to build TF. Once the request is received, the system will check the cache to see if we already have such a package built. If yes, then it will send back the URL for downloading this package. A sample proto (not final) is shown below:
+
+```
+proto build_options {
+   string version = 0;
+   enum CpuOptions {
+       SSE3 = 0;
+       SSE4 = 1;
+       SSE4_1 = 2;
+       SSE4_A = 3;
+       SSE4_2 = 4;
+       AVX = 5;
+       AVX2 = 6;
+       AVX512F = 7;
+   }
+   repeated CpuOptions cpu_options = 1;
+
+   // CUDA options
+   bool cuda_enabled = 2;
+   enum NvidiaGPUGeneration {
+       KEPLER = 0;
+       MAXWELL = 1;
+       PASCAL = 2;
+       VOLTA = 3;
+       TURING = 4;
+   }
+   NvidiaGPUGeneration gpu_generation = 3;
+   string CUDA_version = 4;
+   string cuDNN version = 5;
+……
+   // Free formed string of options to append to bazel.
+   string extra_options = 100;
+}
+```
+
+If no such package exists in cache, then the system will execute the following commands with the appropriate flags and environment variables to newly build it.
+
+```
+git checkout <tag>
+yes "" | ./configure
+bazel build tensorflow/tools/pip_package:build_pip_package
+./bazel-bin/tensorflow/tools/pip_package/build_pip_package
+```
+
+Once the build is complete, the package will be stored in the cache before sending out a link to the user for download.
+
+**Package Cache**
+
+We will design the package cache as a simple GCS bucket supported by a simple database. The database can have the following schema: 
+
+```
+PackageCacheDataStore {
+   // This is the primary key to our datastore.
+   // As build_options is such a complex data type,
+   // this will be the the build_options fed through a hash function
+   string build_options_key;
+
+   // The path to the artifact. Will look like:
+   // gs://tensorflow-artifacts/on-demand/<hash>/packagename.whl
+   string artifact_location;
+
+   // Maybe, maybe not
+   date last_built;
+   date last_downloaded;
+   int download_count;
+}
+```
+
+### Other Details
+
+**Bug Filing**
```

Comment:
What is the plan to allow feature requests/PR submissions? For example, how would I request for the system to build a binary with compiler flags for the latest Xeon instruction set?

Reply:
And how frequently will the build system compiler be updated to support updated `-m` flags?

Reply:
As long the compiler flags are available for the OC+gcc combination it can be passed in to the build environment.

Reply:
`OC`?

Reply:
I meant OS(operating System)